node/src/components/consensus/protocols/highway/performance_meter.rs

```diff
@@ -0,0 +1,137 @@
+use std::cmp::{max, min};
+
+use datasize::DataSize;
+use tracing::error;
+
+use casper_types::{TimeDiff, Timestamp};
+
+use crate::components::consensus::{
+    highway_core::{
+        finality_detector::{
+            assigned_weight_and_latest_unit, find_max_quora, round_participation,
+            RoundParticipation,
+        },
+        state, State,
+    },
+    traits::Context,
+    utils::ValidatorIndex,
+};
+
+/// The number of most recent blocks for which we average the max quorum in which we participated.
+const BLOCKS_TO_CONSIDER: usize = 5;
+/// The average max quorum that triggers us to slow down: with this big or smaller average max
+/// quorum per `BLOCKS_TO_CONSIDER`, we increase our round length.
+const SLOW_DOWN_THRESHOLD: f64 = 0.8;
+/// The average max quorum that triggers us to speed up: with this big or larger average max quorum
+/// per `BLOCKS_TO_CONSIDER`, we decrease our round length.
+const ACCELERATION_THRESHOLD: f64 = 0.9;
+/// We will try to accelerate (decrease our round length) every `ACCELERATION_PARAMETER` rounds if
+/// we have a big enough average max quorum.
+const ACCELERATION_PARAMETER: u64 = 40;
+
+#[derive(DataSize, Debug, Clone)]
+pub(crate) struct PerformanceMeter {
+    our_vid: ValidatorIndex,
+    min_round_len: TimeDiff,
+    max_round_len: TimeDiff,
+    current_round_len: TimeDiff,
+    last_switch_round_id: Timestamp,
```

Comment:
I think this doesn't have anything to do with switch blocks, so maybe it could be renamed to `last_changed_round_id` or similar.  Or even a comment on the variable would help if you prefer to not rename it?